src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -2902,12 +2886,29 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           }
         }
 
-        // if we had missing param types, see if we can undo eta-expansion and recover type info
-        val expectedFunTypeBeforeEtaExpansion =
-          if (paramsMissingType.isEmpty) NoType
-          else typeUnEtaExpanded
+        // If we're typing `(a1: T1, ..., aN: TN) => m(a1,..., aN)`, where some Ti are not fully defined,
+        // type `m` directly (undoing eta-expansion of method m) to determine the argument types.
+        val ptUnrollingEtaExpansion =
+          if (paramsMissingType.nonEmpty && pt != ErrorType) fun.body match {
+            case Apply(meth, args) if (vparams corresponds args) { case (p, Ident(name)) => p.name == name case _ => false } =>
+              val methArgs = NoSymbol.newSyntheticValueParams(argpts map { case NoType => WildcardType case tp => tp })
+              // we're looking for a method (as indicated by FUNmode), so let's make sure our expected type is a MethodType
+              val methPt = MethodType(methArgs, respt)
+
+              silent(_.typed(meth, mode.forFunMode, methPt)) filter (_ => context.undetparams.isEmpty) map { methTyped =>
+                // if context.undetparams is not empty, the method was polymorphic,
+                // so we need the missing arguments to infer its type. See #871
+                val funPt = normalize(methTyped.tpe) baseType FunctionClass(numVparams)
```

Comment:
Considering the overloaded case. AFAICT, `normalize` would return an `OverloadedType`, for which `baseType` would give `NoType`, which would bail out here.

The status quo (shown below) doesn't seem to try to do any overload resolution (even just on arity). 

```
Welcome to Scala 2.11.8 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_71).
Type in expressions for evaluation. Or try :help.

scala> object NonOL { def foo(a: Int) = a }
defined object NonOL

scala> (a => NonOL.foo(a))
res0: Int => Int = <function1>

scala> object OL1 { def foo(a: Int) = a; def foo(a: Any, b: Any) = a }
defined object OL1

scala> ((a) => OL1.foo(a))
<console>:13: error: missing parameter type
       ((a) => OL1.foo(a))
         ^

scala> ((a, b) => OL1.foo(a, b))
<console>:13: error: missing parameter type
       ((a, b) => OL1.foo(a, b))
         ^
<console>:13: error: missing parameter type
       ((a, b) => OL1.foo(a, b))
            ^

scala> object OL2 { def foo(a: Int) = a; def foo(a: Any) = a }
defined object OL2

scala> ((a) => OL2.foo(a))
<console>:13: error: missing parameter type
       ((a) => OL2.foo(a))
         ^
```